### PR TITLE
[FLINK-8139][table] Check for proper equals() and hashCode() when reg…

### PR DESCRIPTION
**Pull request by Aegeaner**
https://issues.apache.org/jira/browse/FLINK-8139

Check for proper equals() and hashCode() when registering a table.

## Verifying this change

flink-table unit tests passed.

### CODE REVIEW

**Comment by fhueske** [comment]
Hi @Aegeaner, thanks for the pull request. 
I had a brief look at it and noticed a few things:

1. Please fill out the PR template (see #5063 as an example)
2. Please add unit tests that validate that the checks you added are working correctly. 
3. Please update the `equals()` and `hashCode()` methods of `Row` as mentioned in the JIRA issue.

Thank you, Fabian

**Comment by Aegeaner** [comment]
Hi @fhueske , thanks for review. 
I will update` equals() `and `hashcode() `methods of Row, but I also noticed that in `JoinOperator` and `KeySelector`, they used `RowComparator` to compare row elements. 
For the unit test, I noticed in `TypeInformationTestBase` there is a` testHashcodeAndEquals` method already, I will complete this test to add test case with type info without implementing its own `hashcode ` and `equals` method.

**Comment by Aegeaner** [comment]
@fhueske I found `TypeInformation `class has abstract `equals()` and `hashcode()`  override methods defined, so any type info must implement these methods. So is reflection check necessary?

**Comment by twalthr** [comment]
@Aegeaner `equals` and `hashCode` in `TypeInformation` are meant for comparing type information and not instances of values with this type. The `RowComparator` internally uses `hashCode` and `equals` of the objects as well.

**Comment by Aegeaner** [comment]
@twalthr But when registering a table/TableSource, we only got a `TableSchema`, which is an array of `TypeInformation` ? How should I check the instances of values?

**Comment by fhueske** [comment]
Hmm, I agree with @twalthr. Adding these checks when a DataSet/DataStream is imported would prevent valid queries (e.g., those that to simple transformations). I would not add a config switch to disable the global checks. Adding the checks to the operators that require properly implemented `equals()` and `hashCode()` methods sounds like a better approach. 

We should also add to the documentation that all objects that are processed by the Table API / SQL should have equals and hashCode properly implemented.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala [hunk 1]
```diff
@@ -98,7 +98,9 @@ abstract class BatchTableEnvironment(
 
     tableSource match {
       case batchTableSource: BatchTableSource[_] =>
-        registerTableInternal(name, new BatchTableSourceTable(batchTableSource))
+        val table = new BatchTableSourceTable(batchTableSource)
+        checkValidTableSourceType(tableSource)
+        registerTableInternal(name, table)
```
[the remark this answers is not in the document]
Comment (Aegeaner) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala]: `registerTableInternal` takes parameter with `AbstractTable` type, while when registering `TableSource`, we should pass `TableSource` type.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala [hunk 2]
```diff
@@ -411,6 +411,7 @@ abstract class TableEnvironment(val config: TableConfig) {
     }
 
     checkValidTableName(name)
+    checkValidTableType(table)
```
[the remark this answers is not in the document]
Comment (Aegeaner) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: `org.apache.flink.table.api.TableEnvironment#validateType` is not only used to validate output types, but also used for all validating type information case, e.g. `BatchTableEnvironment#translate` for explain statement.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala [hunk 3]
```diff
@@ -721,6 +722,42 @@ abstract class TableEnvironment(val config: TableConfig) {
     */
   protected def checkValidTableName(name: String): Unit
 
+  /**
+    * Checks if the chosen table type is valid.
+    * @param table The table to check
+    */
+  protected def checkValidTableType(table: Table): Unit = {
```
[the remark this answers is not in the document]
Comment (Aegeaner) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: I will do this.

[the remark this answers is not in the document]
Comment (Aegeaner) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: I will do this

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala [hunk 4]
```diff
@@ -721,6 +722,42 @@ abstract class TableEnvironment(val config: TableConfig) {
...
+    val types = table.getSchema.getTypes
+    checkTypeArray(types)
+  }
+
+  private def checkTypeArray(types: Array[TypeInformation[_]]) = {
+    for (typeInfo <- types) {
+      if(!typeInfo.asInstanceOf[TypeInformation[_]].isBasicType &&
+        !typeInfo.isInstanceOf[PrimitiveArrayTypeInfo[_]]) {
```
[the remark this answers is not in the document]
Comment (Aegeaner) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: I will do this.

**Comment by twalthr** [comment]
Thank you @Aegeaner. I will merge this...

**Pull request closed without merging**